File: src/wiki/schemas/site.py
"""Site build models for virtual pages and table of contents."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from .wiki_config import Config


class TocItem(BaseModel):
    model_config = ConfigDict(extra="ignore")

    title: str
    slug: str
    level: int


class VirtualPage(BaseModel):
    model_config = ConfigDict(extra="ignore")

    file_slug: str
    title: str
    markdown: str
    html: str
    frontmatter: dict[str, Any]
    layout_path: Path | None = None
    layout_stem: str = "default"
    wiki_ids: list[str] = Field(default_factory=list)
    outline: list[TocItem] = Field(default_factory=list)
    backlink_slugs: list[str] = Field(default_factory=list)

    @property
    def full_slug(self) -> str:
        return self.file_slug

    @property
    def has_frontmatter(self) -> bool:
        return bool(self.frontmatter)


class WikiSite(BaseModel):
    model_config = ConfigDict(extra="ignore", arbitrary_types_allowed=True)

    pages: list[VirtualPage]
    config: Config | None = None
    pages_by_route: dict[str, VirtualPage] = Field(default_factory=dict)
    routes_by_wiki_id: dict[str, str] = Field(default_factory=dict)

File: src/wiki/schemas/site.ts
import type { Config } from "./wiki_config.ts";

export interface TocItem {
  readonly title: string;
  readonly slug: string;
  readonly level: number;
}

export interface VirtualPage {
  readonly file_slug: string;
  readonly title: string;
  readonly markdown: string;
  readonly html: string;
  readonly frontmatter: Record<string, unknown>;
  readonly source_path: string | null;
  readonly layout_path: string | null;
  readonly layout_stem: string;
  readonly wiki_ids: readonly string[];
  readonly outline: readonly TocItem[];
  readonly backlink_slugs: readonly string[];
}

export interface WikiSite {
  readonly pages: readonly VirtualPage[];
  readonly config: Config;
  readonly pages_by_route: ReadonlyMap<string, VirtualPage>;
  readonly routes_by_wiki_id: ReadonlyMap<string, string>;
}
